test-common/src/main/scala/org/scalajs/testcommon/RPCCore.scala

```diff
@@ -0,0 +1,187 @@
+    Serializer.withInputStream(msg) { in =>
+      val opCode = in.readByte()
+      val callId = in.readLong()
+
+      opCode match {
+        case RPCCore.REPLY_OK =>
+          Option(pending.remove(callId)).foreach { p =>
+            import p._
+            promise.complete(Try(deserialize[Resp](in)))
+          }
+
+        case RPCCore.REPLY_ERR =>
+          Option(pending.remove(callId)).foreach { p =>
+            val throwable = Try(deserialize[Throwable](in)) match {
+              case Success(t) => new RPCException(t)
+              case Failure(t) => t
+            }
+
+            p.promise.failure(throwable)
+          }
+
+        case _ =>
+          Option(endpoints.get(opCode)).fold {
+            // Failure[Unit] to not confuse serialization implicits.
+            val repl = Failure[Unit](new IllegalArgumentException(
+                s"Unknown opcode: $opCode"))
+            send(makeReply(callId, repl))
+          } { bep =>
+            val ep: bep.endpoint.type = bep.endpoint
+            import ep._
+
+            import scala.concurrent.ExecutionContext.Implicits.global
+
+            futureFromTry(Try(deserialize[ep.Req](in)))
+              .flatMap(bep.exec)
+              .onComplete(repl => send(makeReply(callId, repl)))
+          }
+      }
+    }
+  }
+
+  /** Subclass needs to implement message sending. */
+  protected def send(msg: String): Unit
+
+  /** Used to make an actual call to the other end. */
+  final def call(ep: RPCEndpoint)(req: ep.Req): Future[ep.Resp] = {
+    import ep._
+
+    // Reserve an id for this call.
+    val id = nextId.incrementAndGet()
+
+    // Prepare message.
+    val msg = makeMsg(opCode, id, req)
+
+    // Register pending call.
+    val promise = Promise[Resp]
+    val oldCall = pending.put(id, PendingCall(promise))
+
+    if (oldCall != null) {
+      throw new AssertionError("Ran out of call ids!")
+    }
+
+    // Actually send message.
+    send(msg)
+
+    promise.future
+  }
+
+  /** Attaches the given method to the given (local) endpoint. */
+  final def attach(ep: RPCEndpoint)(ex: ep.Req => ep.Resp): Unit = {
+    attachAsync(ep)(x => futureFromTry(Try(ex(x))))
+  }
+
+  /** Attaches the given method to the given (local) endpoint. */
+  final def attachAsync(ep: RPCEndpoint)(ex: ep.Req => Future[ep.Resp]): Unit = {
+    require(ep.opCode != REPLY_OK && ep.opCode != REPLY_ERR, "Reserved op code")
+
+    val bound = new BoundEndpoint {
+      val endpoint: ep.type = ep
+      val exec = ex
+    }
+
+    val old = endpoints.put(ep.opCode, bound)
+    require(old == null, "Endpoint already was attached.")
+  }
+
+  final def detach(ep: RPCEndpoint): Unit = {
+    val old = endpoints.remove(ep.opCode)
+    require(old != null, "Endpoint was not attached.")
+  }
+
+  /** Close the communication channel. */
+  def close(): Unit = {
+    for {
+      (callId, p) <- pending.asScala
+      failing <- Option(pending.remove(callId))
```

Comment:
I'm using `keySet` which is guaranteed to be safe for this kind of operations: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentHashMap.html#keySet()